File: Sources/GatewayCore/OllamaModelStore.swift
import Foundation

/// Reads Ollama's local OCI-style model store without requiring its server to
/// be running. A manifest is usable only when every descriptor resolves to a
/// blob of the declared size.
public enum OllamaModelStore {
    private struct Manifest: Decodable {
        var schemaVersion: Int
        var config: Descriptor
        var layers: [Descriptor]
    }

    private struct Descriptor: Decodable {
        var digest: String
        var size: Int64
    }

    public static func hasCompleteModel(_ name: String, modelsRoot: URL,
                                        fileManager: FileManager = .default) -> Bool {
        guard let manifestURL = manifestURL(for: name, modelsRoot: modelsRoot),
              let data = try? Data(contentsOf: manifestURL),
              let manifest = try? JSONDecoder().decode(Manifest.self, from: data),
              manifest.schemaVersion == 2,
              !manifest.layers.isEmpty
        else { return false }

        return ([manifest.config] + manifest.layers).allSatisfy { descriptor in
            guard descriptor.size >= 0,
                  let blobName = blobName(for: descriptor.digest)
            else { return false }
            let blob = modelsRoot.appending(path: "blobs").appending(path: blobName)
                .resolvingSymlinksInPath()
            guard let attributes = try? fileManager.attributesOfItem(atPath: blob.path),
                  let bytes = attributes[.size] as? NSNumber
            else { return false }
            return bytes.int64Value == descriptor.size
        }
    }

    public static func hasManifest(_ name: String, modelsRoot: URL,
                                   fileManager: FileManager = .default) -> Bool {
        guard let manifest = manifestURL(for: name, modelsRoot: modelsRoot) else { return false }
        return fileManager.fileExists(atPath: manifest.path)
    }

    private static func manifestURL(for name: String, modelsRoot: URL) -> URL? {
        let nameAndTag = name.split(separator: ":", maxSplits: 1,
                                    omittingEmptySubsequences: false).map(String.init)
        guard !nameAndTag[0].isEmpty else { return nil }
        let tag = nameAndTag.count == 2 ? nameAndTag[1] : "latest"
        guard safeComponent(tag) else { return nil }

        let path = nameAndTag[0].split(separator: "/",
                                      omittingEmptySubsequences: false).map(String.init)
        guard !path.isEmpty, path.allSatisfy(safeComponent) else { return nil }
        let namespace = path.count == 1 ? "library" : path.dropLast().joined(separator: "/")
        let model = path.last!
        return modelsRoot.appending(path: "manifests/registry.ollama.ai")
            .appending(path: namespace).appending(path: model).appending(path: tag)
    }

    private static func blobName(for digest: String) -> String? {
        let parts = digest.split(separator: ":", omittingEmptySubsequences: false)
        guard parts.count == 2, parts[0] == "sha256", parts[1].count == 64,
              parts[1].allSatisfy({ $0.isHexDigit })
        else { return nil }
        return "sha256-\(parts[1])"
    }

    private static func safeComponent(_ value: String) -> Bool {
        !value.isEmpty && value != "." && value != ".."
            && !value.contains("/") && !value.contains("\\")
    }
}

File: cross-platform/src/core/ollamaModelStore.ts
/**
 * Reads Ollama's local OCI-style model store without requiring its server
 * to be running, ported from `OllamaModelStore.swift`. A manifest is usable
 * only when every descriptor resolves to a blob of the declared size.
 *
 * Ollama's own on-disk layout is identical on every host platform — this is
 * not Apple-specific, unlike `OllamaRelease.swift` (a pinned macOS `.dmg`
 * installer with an Apple Team ID, deliberately not ported here).
 */
import { existsSync, readFileSync, realpathSync, statSync } from "fs";
import { join } from "path";

interface Descriptor { digest: string; size: number }
interface Manifest { schemaVersion: number; config: Descriptor; layers: Descriptor[] }

function safeComponent(value: string): boolean {
  return value !== "" && value !== "." && value !== ".."
    && !value.includes("/") && !value.includes("\\");
}

function blobName(digest: string): string | undefined {
  const parts = digest.split(":");
  if (parts.length !== 2 || parts[0] !== "sha256") return undefined;
  const hex = parts[1]!;
  if (hex.length !== 64 || !/^[0-9a-fA-F]+$/.test(hex)) return undefined;
  return `sha256-${hex}`;
}

function manifestPath(name: string, modelsRoot: string): string | undefined {
  // Swift's `split(separator:maxSplits:omittingEmptySubsequences:false)`
  // keeps empty pieces and caps at one split — "a:b:c" becomes ["a","b:c"],
  // and a name with no colon becomes ["name"] (no second element), not
  // ["name", ""].
  const colon = name.indexOf(":");
  const nameAndTag = colon < 0 ? [name] : [name.slice(0, colon), name.slice(colon + 1)];
  if (nameAndTag[0] === "") return undefined;
  const tag = nameAndTag.length === 2 ? nameAndTag[1]! : "latest";
  if (!safeComponent(tag)) return undefined;

  const path = nameAndTag[0]!.split("/");
  if (path.length === 0 || !path.every(safeComponent)) return undefined;
  const namespace = path.length === 1 ? "library" : path.slice(0, -1).join("/");
  const model = path[path.length - 1]!;
  return join(modelsRoot, "manifests", "registry.ollama.ai", namespace, model, tag);
}

export function hasManifest(name: string, modelsRoot: string): boolean {
  const p = manifestPath(name, modelsRoot);
  return p !== undefined && existsSync(p);
}

export function hasCompleteModel(name: string, modelsRoot: string): boolean {
  const p = manifestPath(name, modelsRoot);
  if (p === undefined) return false;
  let manifest: Manifest;
  try {
    const raw = JSON.parse(readFileSync(p, "utf8")) as Record<string, unknown>;
    if (typeof raw.schemaVersion !== "number" || raw.schemaVersion !== 2) return false;
    const config = raw.config as Record<string, unknown>;
    const layersRaw = raw.layers;
    if (!Array.isArray(layersRaw) || layersRaw.length === 0) return false;
    if (typeof config?.digest !== "string" || typeof config?.size !== "number") return false;
    const layers: Descriptor[] = layersRaw.map(l => {
      const r = l as Record<string, unknown>;
      if (typeof r.digest !== "string" || typeof r.size !== "number") throw new Error("bad layer");
      return { digest: r.digest, size: r.size };
    });
    manifest = { schemaVersion: raw.schemaVersion, config: { digest: config.digest, size: config.size }, layers };
  } catch { return false; }

  const descriptors = [manifest.config, ...manifest.layers];
  return descriptors.every(d => {
    if (!(d.size >= 0)) return false;
    const name = blobName(d.digest);
    if (name === undefined) return false;
    const blob = join(modelsRoot, "blobs", name);
    let real: string;
    try { real = realpathSync(blob); } catch { return false; }
    let size: number;
    try { size = statSync(real).size; } catch { return false; }
    return size === d.size;
  });
}
